### ISSUE TITLE
bowtie2 sam output format error 

### ISSUE THREAD

**Issue by nick-youngblut** [BenLangmead/bowtie2#155]
I'm running bowtie2 v2.3.3.1 installed via conda from the bioconda channel (Ubuntu 16.04.3; NFS file system). 

bowtie2 stochastically generates incorrectly formatted SAM files. In the SAM snippet below, a line break is missing from the sam file.  It appears that the problem is an issue with multiple threads writing to the same file. Maybe a buffering issue? 

Any ideas on what I can do to prevent these SAM formatting errors?

```
D00611:353:HWWTLBCXX:2:1116:4683:22706	77	*	0	0	*	*	0	0	GTCTGAACTCCAGTCACAAGCGTACGTCCATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAAGACTTGCTGCTGCCTATATGCCACCGCAGTACAAACTGCTGAGTACGCGATGTGAGACTGTTGGAACCTGTTACGTCCACTTCCTCTTGTTTGCGCTCAGACCATCTTAAAATGAACCGAATCAGCGGTGTATCTTATGGTCACAAGTAAATATGCTTGGTGATGTTCTATGTGCATGGAGACTG	GAAAAAAGAGGGIGGIGGIIIGIGGGGAGG.<GGIIIGGGGG.AG..AGGGAAGGGGI<<A.<<<....<........<<.<..<....<<..........<............<.......<<...........<....<..<.......<......<.<......<..<...<.G..<......<...<7...<.......<......<......7...<...7<.<.....7.<..<.....<....	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:4683:22706	141	*	0	0	*	*	0	0	TGAGCTGCTCAAAGAAGTAGCTTTTGATTCTCCACGGAATCGTCATCAGAATGGTGCGGCGGGTGGGACTGGCATAGCATCTGCCTCGGCTTGATGCTCGGTTACGGCCCTTGCCTATACTTCTCACAATACGTACTTCACCACTCTTTCAAGATCTAGTGGGTAAGTCCGTCGTGATGCGTTGATGTGCAGGTCGATCGTTAGTACTAAGGTGATGACTTCTTTGTGGGTTATTATATATTGCTCAGTT	.<.........<<......<....................................................<....<.<..<.<....<.......<...<........<......<.....<.......<<<...<...<.<<..<....<.......<.7.....<.......7...7.............................<7..7..........7..7....................7	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:10838:22681	77	*	0	0	*	*	0	0	GGTGACGGTGATAGAAGCGAGGGTGCCGGTGAATGTTCTCATCTCCAGCCATTTTAAAGCGGTAAAAAGCATTGATACATTCACAGCCCACACTGAAAATCCGTGGGGGATTTGAAACAGATAATGCAATAGCACCGCAATACCTGCTGCACCGCCTGTCGGATTATCATGCGGAAATTAAAACAAGGACAGATCGGAAGAGCACACGTCTGAACTCCAGTCACAAGCGTACGTCCAAGTCGTATGCGGG	AGGGAGA.GAGGGGGGGIIGGG.<<GGAGAGAAAGGAGGGGAGGGGGGGAGAAGGGGA.<<GG...<.GAGGAAG.GGG<<GGIGGGGGAAGG<..<<.<G<<<<.<GGG.GG.<<GGGGAGA<G.GGAG.<GGGGGAGG<<AAAAGAGGAAGGI<AG<AGG<<<G..<.<....<<<.<.<..<.<<GGAA.7.<<.7<.77..<A<<GAA.<<.<<.<<.<<7AAGGGGAAAG....7.7...<7.<.	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:10838:22681	141	*	0	0	*	*	0	0	GTCCTTGTTTTTATTTCCGCATGATATTCCGACAGGCGGTGCAGCAGGTATTGCGGTGCTATTGCATTATCTGTTTCAAATCCCCCACGGAGTTTCAGTGTGGGCTGTGAATGTATCAATGCTTTTTACCGCTTTAAAATGGCTGGAGATGAGAACATTCACCGGCACCCTCGCTTGTATCACCGGCACCAGATCGGAAGAGCGTCGCGGAGGGAGAGAGCGTAGGTGGCCGTGGGGGGCCGGCCCTGGA	AGGAAGAGAAGGGGIIGAAGAG.<<<AA<AG.AA.GGGIIG.GG..GGGGGA.G..GGAAGGGGAG<G<<<GAGGGG.<GGAAGGGAGGG..<<AAAAA<<<GAGAGG.GAGGA..A.<GA.AGA<GAA<.A<.<<GGAAGIGGGAGG.<..GGGGGAGGGGGGGIG.<GGGAG...<<AG.7GA..AA<.<.<<<A..777AG7.....7<GG<.7..7.7...77....7.<...77.....<.....	YT:Z:UD00611:353:HWWTLBCXX:2:1116:18804:22057	77	*	0	0	*	*	0	0	GATCGGAAGAGCACACGTCTGAACTCCAGTCACAAGCGTAGGTCCATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAGCACACGTCTGAACTCCAGTCACAAGCGTACGTCCATCTCGTATGCCGTCTTCTGCTTGTAACAGAAAAGCGCCCGTCACAACTCAAGAAACAAACGTACGGGAGACTCGAATGAATACTCATGCAATACCAAACCAAGGAAAGATCACGCACATACGCAAAGCATAAGTA	GGGGGIGGGGIIIIIIGIIIIIIIIIIIIIIIIAG..<<..<GGIIGIIIIIIIIIIIIGIIIIIIIIGAGG...<A<G..<A.AA.AA..GA<AGAG.<<<GAA..<<A..<.G<.<GGAGGAGAG.<<...<G<.<..<.<.<<.........<........<.........<.<..........<<.............<....<..<........<<.....7........<......7.......	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:18804:22057	141	*	0	0	*	*	0	0	GATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGAATCATTAAAAAAAAAAAAAATCGGAAGAGCGCAGACAAGAGGAAAAGAGGCCGGTGCCTGCGGTGCCCGCACATGCACAACGAACCCGAATGGACGCGAACACAGAAGCTCCGCCATACAGCGCAACCTCAGCGCGACCCATCATGCCCCGTATCACACGTTCACCTACTGAGCGACTACCGAGCCGCCAAACTCCCAT	.GAGAAGGIIGGGAA<AAGGIGGAAGGGGGAAGAA<<<GA.GGA..<GGG...<<G<.<<A<.<<AA.<.A.<..AA<G.<.<..............<....<.........<<..<<...<..<.<<.<.<........<...<....<..G.<.......7......<.<....7...7..7.77..<.7........<..7<...7.........7.<..7.77..7..7....7.<.....7..7.	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:7906:22497	77	*	0	0	*	*	0	0	GATCGGAAGAGCACACGTCTGAACTCCAGTCACAAGCGTACGTCCATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAGCACACGTCTGTCCGCCCGTCACAAGCGTCCGTCGATCAGGTGGCCCTCATAATAAGGGAAAAATCAAACAGATATCACGCGCACACGCCGTAAATGCCGTCACGCGCTCTGTTCCACCTAGAATGCCGCATGTTGCTTGCAACAAAACTATACTGAAATGACGACGGCT	AAGAG<GGGGGGAGGGGIIGAAGGGGGGGIGIIIGIIAGIGGIGGG<GGGGGAA..<.GA.<<A...<<..<<<<.AG......<<.<.<...<.......<..<GAA...A....<....<.<.<..<<<..<..<.<<..<.<...G.A<....<<...<<<<.<....<<<.....<...<.<.<..<<....<..............7.....<.<<<......7.......<...<7........	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:7906:22497	141	*	0	0	*	*	0	0	GATCGGAAGAGCGCCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATTAAAAAAAAAAAAAAATACACCTCGCGTACTGACGATAACGCAACTAGCCCACTCTATCCGATTGGCTATGGATTGAACTCCACCACTTTCAGCAGATCGGAAGAGAGTCGTGTAGTAACAGAATGGTCATACACGCGGCAGCCACGACATACCAACCAAACAACGCACGCGGTCGTGAGCGCGCCAACGCGAC	AAGGGGGIGIGIG.<GGGGGGGAGGAG.AGAGA.GGAGGGGGGG<A.AG..<.<.<G.<GGGG..<GG.<<...<<.....<.<..<.<.<.<<..<..<.<.<A..<<GGA.<.AA.<.....<<..G.<.<......<<.A..GG..G.<.<.<..7<7..<..<....<....77...<....7.........7....7..7<....7.....<..............7..7....<..7.<77...	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:18996:22070	77	*	0	0	*	*	0	0	GCAGACCGATACAGATTCGTAGCGATCGAAATGGCTTCCAGGTTGAGTGACTCGACGGGCATTTGTTTGAGCTTAGAAAGTGTATCGTGTTCTTCCTTCCGAAGCTGCTGATGTTGATGTGAACGTTCCATTTTTTAGGCCTGCCTTAAATTATCAGAATCATTAGCTATGAAACGAATTTTAACAGAGTCTGGTGTGAACTTCAAGCCCATATATCGGTTGAATAAAAAAAGCCCAAACAACGCCAAAA	GGGGGGGGIIIIIIGGIIIGGGGGGGIIIIIIIIIGGGGGGGGGGGGGGGIGGGIIIGGGIGGIIIIIGIIIIGIGIIIGGGIG<GGGGGIIGIIIGGGGGGGGIIIGGGAGGGIIIGGGGGGIIGGGGIIGGGIGIIIGGGIGGGIGIIIIIGIGGGGGGGIGIIIIIIIGIIIGGIIIIIGIGIAGIGGGGGGGGGGIIIIIAGGGIIGIIA.GGAAAGAGAAAGGAGIIGIGGG.G.77AA.A7AGG	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:18996:22070	141	*	0	0	*	*	0	0	ACAAAAACAACAGAATGGCCAAAAGGGAGATAATCGCTAACGTCACTACATGCTCTGTCGAAAACAGATGAAACGGATCATGTTTATAGTCGGATTGAACATATTTTTGCAAAACTGCACCCCCGTTTTCGGGTTTTTGTTTATTTTGGCGTTGTTTGGGCTTTTTTTATTCAACCGATATATGGTCTTGAAGTTCACACCAGACTCTGTTAAAATTCGTTTCATAGCTAATGATTCTGATAATTTAAGG	GGAGAGGIGIIAGAGGGGAAGIGGGGGGIGGGGGGIIIIII<GGGGGIIIIIGGIIIAGGGGGGGGGIGIIIIGGGIGIGGIIIGGGGI<GGGGAGIGGGGGGGGAGGIIGGGIIIIIIIGGGGG<GGGGAGGGGGIGGGIGGGGGGAAGGAGGGAGGGGGGGIGGGIIGGGGIGGGGGAAAGGG.AGAAGAG<AGAGGAA7A.GAG.A7GGGA...7A.A.777AGGG.A7<..7777A.7AAAA7A.7	YT:Z:UP
D00611:353:HWWTLBCXX:2:1116:8466:22309	77	*	0	0	*	*	0	0	TACGCGTATCCTGAAAAGGGAGTTGCTTTTGCCAGAATTCCATATTTTCAAGATGGGAGCATTATGAGTTTTGATTGTCTGTTTGCGGTTAACGATGAAAAGATGCGTGCATTTCTGGAGGGAGTCAGACCGCGTCTTTGGGAAAAAAGCAAGCGGAAAGTGACCGTGTTTACGGATGGAGATGGAGGGACTTCAAGAGAGCAGGAAGCCATTGTCAGAGAGGTTCAGCGGAGTCAAGTCATCATGAATC	GGGGGGGAGGGGIG<AAAGGGGAGG<<GGGGG<GGGGGIIGGGGIGIIGGGGG<GGGAGAGGG.GG.AAGGGGGGGIIGIGGGGIGGIIIIGGIAGGGIIGGGGGGGIIIGGIIIIIG<GGGGGGGGIAGGGGIGIIGGGIGGGIGIIIIGGGGGGGGGGIIGGIGGGGGGIIGG.GGG.<<AGG.GGG<<GGGGG<7<AGGGGGAGAAAGGGGGAG<<<7<A77<<<.7<<<<<AGG7.<AGGG<<7AA	YT:Z:UP
```

**Comment by nick-youngblut** [comment]
To clarify the issue, this SAM formatting issue seems to be a synchronization issue, and it only occurs on the SGE compute cluster that I'm using, but doesn't occur when `bowtie2` is run on a headless server (both use the same Ubuntu OS and file system). 

The compute cluster admin thinks the problem is caused by the synchronization primitive that's used by bowtie2. Which synchronization primitive is used by `bowtie2`: POSIX locks, flock()
or semaphores?

**Comment by BenLangmead** [comment]
It uses TBB which in turn uses pthreads.  To the best of my knowledge Pthreads in turn uses the Linux "futex" mechanism.

**Comment by BenLangmead** [comment]
Note that we have a potential fix for this issue.  If you have the chance, there is an alpha build with a potential fix linked to from issue #158.  If you can try it and let us know if the problem is fixed, we'd appreciate it.  Either way, a new release with that fix is coming very soon.

**Comment by nick-youngblut** [comment]
Thanks for the letting me know! I've had a hard time consistently recreating the issue on a small dataset. Usually I have to run bowtie2 on many large read files in order to have >=1 job generate an error.  Regardless, I'll try the alpha build and see if I get any errors.

**Comment by ch4rr0** [comment]
Based on feedback we received from others we think this already fixed in v2.3.4.1. Please feel free to reopen if you still see it.
